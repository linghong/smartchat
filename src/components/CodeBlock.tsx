import React, { useState, useRef, useEffect } from 'react';
import SyntaxHighlighter from 'react-syntax-highlighter/dist/cjs/prism';
import { vscDarkPlus } from 'react-syntax-highlighter/dist/cjs/styles/prism';
import { RiFileCopyLine, RiCheckLine, RiTextWrap } from 'react-icons/ri';

interface CodeBlockProps {
  code: string;
  language: string;
}

const CodeBlock: React.FC<CodeBlockProps> = ({ code, language }) => {
  const [copied, setCopied] = useState(false);
  const [expanded, setExpanded] = useState(false);
  const [wrapLine, setWrapLine] = useState(true);

  const handleCopy = () => {
    navigator.clipboard.writeText(code).then(() => {
      setCopied(true);
      setTimeout(() => setCopied(false), 2000);
    });
  };

  const toggleExpand = () => {
    setWrapLine(!wrapLine);
    setExpanded(!expanded);
  };

  return (
    <div className="relative border border-gray-200 rounded-md">
      <div className="flex justify-between items-center bg-gray-100 px-4 py-2 rounded-t-md">
        <span className="text-sm font-medium text-gray-600">
          {language || 'Code'}
        </span>
        <div className="flex space-x-2">
          <button
            onClick={toggleExpand}
            className="p-1 rounded-md hover:bg-gray-200 transition-colors"
            title={wrapLine ? 'Unwrap lines' : 'Wrap lines'}
          >
            <RiTextWrap size={18} className={wrapLine ? 'text-blue-500' : ''} />
          </button>
          <button
            onClick={handleCopy}
            className="p-1 rounded-md hover:bg-gray-200 transition-colors"
            title="Copy code"
          >
            {copied ? (
              <RiCheckLine size={18} className="text-green-500" />
            ) : (
              <RiFileCopyLine size={18} />
            )}
          </button>
        </div>
      </div>
      <div className={`relative overflow-x-auto`}>
        <SyntaxHighlighter
          language={language}
          style={vscDarkPlus}
          customStyle={{
            padding: '1em',
            fontSize: '14px'
          }}
          wrapLongLines={wrapLine}
        >
          {code}
        </SyntaxHighlighter>
      </div>
    </div>
  );
};

export default CodeBlock;
