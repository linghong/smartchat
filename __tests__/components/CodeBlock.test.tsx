import React from 'react';
import {
  render,
  screen,
  fireEvent,
  waitFor,
  act
} from '@testing-library/react';
import '@testing-library/jest-dom';
import CodeBlock from '@/src/components/CodeBlock';

// Mock the SyntaxHighlighter component
jest.mock('react-syntax-highlighter/dist/cjs/prism', () => {
  const MockSyntaxHighlighter: React.FC<{
    children: string;
    language: string;
    style: any;
    customStyle: React.CSSProperties;
    wrapLongLines: boolean;
  }> = ({ children, language, customStyle, wrapLongLines }) => (
    <pre
      data-testid="mock-syntax-highlighter"
      data-language={language}
      style={customStyle}
      data-wrap-lines={wrapLongLines || undefined}
    >
      {children}
    </pre>
  );
  MockSyntaxHighlighter.displayName = 'MockSyntaxHighlighter';
  return MockSyntaxHighlighter;
});

// Mock clipboard API
const mockClipboard = {
  writeText: jest.fn(() => Promise.resolve())
};
Object.assign(navigator, { clipboard: mockClipboard });

describe('CodeBlock', () => {
  const testCode = 'const greeting = "Hello, World!";';
  const testLanguage = 'javascript';

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
    jest.clearAllMocks();
  });

  it('renders the CodeBlock component', () => {
    render(<CodeBlock code={testCode} language={testLanguage} />);
    const syntaxHighlighter = screen.getByTestId('mock-syntax-highlighter');
    expect(syntaxHighlighter).toBeInTheDocument();
  });

  it('passes the correct props to SyntaxHighlighter', () => {
    render(<CodeBlock code={testCode} language={testLanguage} />);
    const syntaxHighlighter = screen.getByTestId('mock-syntax-highlighter');

    expect(syntaxHighlighter).toHaveAttribute('data-language', testLanguage);
    expect(syntaxHighlighter).toHaveTextContent(testCode);
  });

  it('applies custom styles to SyntaxHighlighter', () => {
    render(<CodeBlock code={testCode} language={testLanguage} />);
    const syntaxHighlighter = screen.getByTestId('mock-syntax-highlighter');

    expect(syntaxHighlighter).toHaveStyle({
      padding: '1em',
      fontSize: '14px',
      maxWidth: '100%',
      overflowX: 'auto'
    });
  });

  it('toggles code wrapping', async () => {
    render(<CodeBlock code={testCode} language={testLanguage} />);
    const syntaxHighlighter = screen.getByTestId('mock-syntax-highlighter');
    const wrapButton = screen.getByTitle('Wrap lines');

    expect(syntaxHighlighter).not.toHaveAttribute('data-wrap-lines');

    await act(async () => {
      fireEvent.click(wrapButton);
    });

    expect(syntaxHighlighter).toHaveAttribute('data-wrap-lines', 'true');

    await act(async () => {
      fireEvent.click(wrapButton);
    });

    expect(syntaxHighlighter).not.toHaveAttribute('data-wrap-lines');
  });

  it('copies code to clipboard and shows confirmation', async () => {
    render(<CodeBlock code={testCode} language={testLanguage} />);
    const copyButton = screen.getByTitle('Copy code');

    // Assert the initial state
    const checkIcon = screen.getByRole('img', { name: /copy code/i });
    expect(checkIcon).not.toHaveClass('text-green-500');

    await act(async () => {
      fireEvent.click(copyButton);
    });

    expect(navigator.clipboard.writeText).toHaveBeenCalledWith(testCode);

    // Assert the state after clicking the "Copy code" button
    const updatedCheckIcon = screen.getByRole('img', { name: /copy code/i });
    expect(updatedCheckIcon).toHaveClass('text-green-500');

    await waitFor(
      () => {
        const finalCheckIcon = screen.getByRole('img', { name: /copy code/i });
        expect(finalCheckIcon).not.toHaveClass('text-green-500');
      },
      { timeout: 3000 }
    );
  });

  it('displays the correct language in the header', () => {
    render(<CodeBlock code={testCode} language={testLanguage} />);
    expect(screen.getByText(testLanguage)).toBeInTheDocument();
  });

  it('displays "Code" in the header when language is not provided', () => {
    render(<CodeBlock code={testCode} language="" />);
    expect(screen.getByText('Code')).toBeInTheDocument();
  });
});
