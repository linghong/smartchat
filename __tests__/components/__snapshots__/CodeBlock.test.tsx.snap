// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeBlock matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="relative border border-gray-200 rounded-md"
  >
    <div
      class="flex justify-between items-center bg-gray-100 px-4 py-2 rounded-t-md"
    >
      <span
        class="text-sm font-medium text-gray-600"
      >
        javascript
      </span>
      <div
        class="flex space-x-2"
      >
        <button
          class="p-1 rounded-md hover:bg-gray-200 transition-colors"
          title="Unwrap lines"
        >
          <svg
            class="text-blue-500"
            fill="currentColor"
            height="18"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 18H16.5C17.8807 18 19 16.8807 19 15.5C19 14.1193 17.8807 13 16.5 13H3V11H16.5C18.9853 11 21 13.0147 21 15.5C21 17.9853 18.9853 20 16.5 20H15V22L11 19L15 16V18ZM3 4H21V6H3V4ZM9 18V20H3V18H9Z"
            />
          </svg>
        </button>
        <button
          class="p-1 rounded-md hover:bg-gray-200 transition-colors"
          title="Copy code"
        >
          <svg
            fill="currentColor"
            height="18"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.9998 6V3C6.9998 2.44772 7.44752 2 7.9998 2H19.9998C20.5521 2 20.9998 2.44772 20.9998 3V17C20.9998 17.5523 20.5521 18 19.9998 18H16.9998V20.9991C16.9998 21.5519 16.5499 22 15.993 22H4.00666C3.45059 22 3 21.5554 3 20.9991L3.0026 7.00087C3.0027 6.44811 3.45264 6 4.00942 6H6.9998ZM5.00242 8L5.00019 20H14.9998V8H5.00242ZM8.9998 6H16.9998V16H18.9998V4H8.9998V6Z"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="relative overflow-x-auto"
    >
      <pre
        data-language="javascript"
        data-testid="mock-syntax-highlighter"
        style="padding: 1em; font-size: 14px;"
      >
        const greeting = "Hello, World!";
      </pre>
    </div>
  </div>
</DocumentFragment>
`;
